File: enigma.cpp
#include <iostream>
#include <string>
#include <unordered_map>
#include <locale>
#include <codecvt>

std::wstring encryptDecrypt(const std::wstring& message, const std::unordered_map<wchar_t, wchar_t>& rotors, const std::unordered_map<wchar_t, wchar_t>& reflector) {
    std::wstring result;
    for (wchar_t charIn : message) {
        if (rotors.count(charIn)) {
            result += rotors.at(charIn);
        } else if (reflector.count(charIn)) {
            result += reflector.at(charIn);
        } else {
            result += charIn; // Оставляем символ, если не найден
        }
    }
    return result;
}

std::unordered_map<wchar_t, wchar_t> invertMap(const std::unordered_map<wchar_t, wchar_t>& m) {
    std::unordered_map<wchar_t, wchar_t> inverted;
    for (const auto& pair : m) {
        inverted[pair.second] = pair.first;
    }
    return inverted;
}

int main() {
    std::unordered_map<wchar_t, wchar_t> rotors = {
        {L'A', L'D'}, {L'B', L'E'}, {L'C', L'F'}, {L'D', L'G'}, {L'E', L'H'}, {L'F', L'I'},
        {L'G', L'J'}, {L'H', L'K'}, {L'I', L'L'}, {L'J', L'M'}, {L'K', L'N'}, {L'L', L'O'},
        {L'M', L'P'}, {L'N', L'Q'}, {L'O', L'R'}, {L'P', L'S'}, {L'Q', L'T'}, {L'R', L'U'},
        {L'S', L'V'}, {L'T', L'W'}, {L'U', L'X'}, {L'V', L'Y'}, {L'W', L'Z'}, {L'X', L'A'},
        {L'Y', L'B'}, {L'Z', L'C'}, {L'А', L'Е'}, {L'Б', L'Ё'}, {L'В', L'Ж'}, {L'Г', L'З'},
        {L'Д', L'И'}, {L'Е', L'Й'}, {L'Ё', L'К'}, {L'Ж', L'Л'}, {L'З', L'М'}, {L'И', L'Н'},
        {L'Й', L'О'}, {L'К', L'П'}, {L'Л', L'Р'}, {L'М', L'С'}, {L'Н', L'Т'}, {L'О', L'У'},
        {L'П', L'Ф'}, {L'Р', L'Х'}, {L'С', L'Ц'}, {L'Т', L'Ч'}, {L'У', L'Ш'}, {L'Ф', L'Щ'},
        {L'Х', L'Ъ'}, {L'Ц', L'Ы'}, {L'Ч', L'Ь'}, {L'Ш', L'Э'}, {L'Щ', L'Ю'}, {L'Ъ', L'Я'},
        {L'Ы', L'А'}, {L'Ь', L'Б'}, {L'Э', L'В'}, {L'Ю', L'Г'}, {L'Я', L'Д'}
    };

    std::unordered_map<wchar_t, wchar_t> reflector = {
        {L'A', L'Y'}, {L'B', L'X'}, {L'C', L'W'}, {L'D', L'V'}, {L'E', L'U'}, {L'F', L'T'},
        {L'G', L'S'}, {L'H', L'R'}, {L'I', L'Q'}, {L'J', L'P'}, {L'K', L'O'}, {L'L', L'N'},
        {L'M', L'M'}, {L'N', L'L'}, {L'O', L'K'}, {L'P', L'J'}, {L'Q', L'I'}, {L'R', L'H'},
        {L'S', L'G'}, {L'T', L'F'}, {L'U', L'E'}, {L'V', L'D'}, {L'W', L'C'}, {L'X', L'B'},
        {L'Y', L'A'}, {L'Z', L'Z'}, {L'А', L'Я'}, {L'Б', L'Ю'}, {L'В', L'Э'}, {L'Г', L'Ь'},
        {L'Д', L'Ы'}, {L'Е', L'Ъ'}, {L'Ё', L'Щ'}, {L'Ж', L'Ш'}, {L'З', L'Ч'}, {L'И', L'Ц'},
        {L'Й', L'Х'}, {L'К', L'Ф'}, {L'Л', L'У'}, {L'М', L'Т'}, {L'Н', L'С'}, {L'О', L'Р'},
        {L'П', L'К'}, {L'Р', L'Й'}, {L'С', L'И'}, {L'Т', L'Г'}, {L'У', L'Е'}, {L'Ф', L'Д'},
        {L'Х', L'В'}, {L'Ц', L'Б'}, {L'Ч', L'А'}, {L'Ш', L'Г'}, {L'Щ', L'Ё'}, {L'Ъ', L'Ж'},
        {L'Ы', L'З'}, {L'Ь', L'Й'}, {L'Э', L'К'}, {L'Ю', L'Л'}, {L'Я', L'М'}
    };

    std::wcin.imbue(std::locale(""));

    while (true) {
        int choice;
        std::wcout << L"Выберите опцию (1 - шифрование, 2 - расшифровка, 0 - выход): ";
        std::wcin >> choice;

        if (choice == 0) {
            std::wcout << L"Выход из программы." << std::endl;
            break;
        }

        std::wstring message;
        std::wcout << L"Введите сообщение: ";
        std::wcin.ignore(); // Очистка буфера перед вводом строки
        std::getline(std::wcin, message);

        for (wchar_t& c : message) {
            c = std::towupper(c); // Приведение к верхнему регистру
        }

        if (choice == 1) {
            std::wstring result = encryptDecrypt(message, rotors, reflector);
            std::wcout << L"Зашифрованное сообщение: " << result << std::endl;
        } else if (choice == 2) {
            auto invertedRotors = invertMap(rotors);
            auto invertedReflector = invertMap(reflector);
            std::wstring result = encryptDecrypt(message, invertedRotors, invertedReflector);
            std::wcout << L"Расшифрованное сообщение: " << result << std::endl;
        } else {
            std::wcout << L"Неверный выбор." << std::endl;
        }
    }

    return 0;
}

File: enigma.py
def encrypt_decrypt(message, rotors, reflector):
    result = []
    for char in message:
        if char in rotors:
            char = rotors[char]
        elif char in reflector:
            char = reflector[char]
        result.append(char)
    return ''.join(result)

def main():
    rotors = {
        'A': 'D', 'B': 'E', 'C': 'F', 'D': 'G', 'E': 'H', 'F': 'I', 'G': 'J',
        'H': 'K', 'I': 'L', 'J': 'M', 'K': 'N', 'L': 'O', 'M': 'P', 'N': 'Q',
        'O': 'R', 'P': 'S', 'Q': 'T', 'R': 'U', 'S': 'V', 'T': 'W', 'U': 'X',
        'V': 'Y', 'W': 'Z', 'X': 'A', 'Y': 'B', 'Z': 'C',
        'А': 'Е', 'Б': 'Ё', 'В': 'Ж', 'Г': 'З', 'Д': 'И', 'Е': 'Й', 'Ё': 'К',
        'Ж': 'Л', 'З': 'М', 'И': 'Н', 'Й': 'О', 'К': 'П', 'Л': 'Р', 'М': 'С',
        'Н': 'Т', 'О': 'У', 'П': 'Ф', 'Р': 'Х', 'С': 'Ц', 'Т': 'Ч', 'У': 'Ш',
        'Ф': 'Щ', 'Х': 'Ъ', 'Ц': 'Ы', 'Ч': 'Ь', 'Ш': 'Э', 'Щ': 'Ю', 'Ъ': 'Я',
        'Ы': 'А', 'Ь': 'Б', 'Э': 'В', 'Ю': 'Г', 'Я': 'Д'
    }

    reflector = {
        'A': 'Y', 'B': 'X', 'C': 'W', 'D': 'V', 'E': 'U', 'F': 'T', 'G': 'S',
        'H': 'R', 'I': 'Q', 'J': 'P', 'K': 'O', 'L': 'N', 'M': 'M', 'N': 'L',
        'O': 'K', 'P': 'J', 'Q': 'I', 'R': 'H', 'S': 'G', 'T': 'F', 'U': 'E',
        'V': 'D', 'W': 'C', 'X': 'B', 'Y': 'A', 'Z': 'Z',
        'А': 'Я', 'Б': 'Ю', 'В': 'Э', 'Г': 'Ь', 'Д': 'Ы', 'Е': 'Ъ', 'Ё': 'Щ',
        'Ж': 'Ш', 'З': 'Ч', 'И': 'Ц', 'Й': 'Х', 'К': 'Ф', 'Л': 'У', 'М': 'Т',
        'Н': 'С', 'О': 'Р', 'П': 'К', 'Р': 'Й', 'С': 'И', 'Т': 'Г', 'У': 'Е',
        'Ф': 'Д', 'Х': 'В', 'Ц': 'Б', 'Ч': 'А', 'Ш': 'Г', 'Щ': 'Ё', 'Ъ': 'Ж',
        'Ы': 'З', 'Ь': 'Й', 'Э': 'К', 'Ю': 'Л', 'Я': 'М'
    }

    while True:
        choice = int(input("Выберите опцию (1 - шифрование, 2 - расшифровка, 0 - выход): "))
        if choice == 0:
            print("Выход из программы.")
            break

        message = input("Введите сообщение: ").upper()

        if choice == 1:
            result = encrypt_decrypt(message, rotors, reflector)
            print("Зашифрованное сообщение:", result)
        elif choice == 2:
            inverted_rotors = {v: k for k, v in rotors.items()}
            inverted_reflector = {v: k for k, v in reflector.items()}
            result = encrypt_decrypt(message, inverted_rotors, inverted_reflector)
            print("Расшифрованное сообщение:", result)
        else:
            print("Неверный выбор.")

if __name__ == "__main__":
    main()
